### ISSUE TITLE
Data Leakage Bug: Model Trained on Full Dataset (Including Val/Test Splits)

### ISSUE THREAD

**Issue by Molten-Ice** [sapientinc/HRM#18]
Hi, first I wanted to say awesome work on the Hierarchical Reasoning Model, super interesting approach, seems especially promising on the Sudoku and Mazes approach.

However, I believe I spotted a data leak in the training and evaluation process for the ARC-AGI dataset. It looks like the model is trained on all of the training dataset and validation dataset example pairs & test pairs, which inflates the metrics unfairly.

I suggest you might want to still train on 300/400 of the validation examples and reserve the 100 others for a completely blind final tests. Test-time training (TTT) could use those reserved 100 example pairs, but never for their test outputs.

The approach seems super promising so it would be great you see how it performs with a distinct train/validation split! Let me know if I can help with a PR or more details - happy to chat!

**Comment by imoneoi** [comment]
We only evaluate on validation set. Initially, we combined train+TTT into one stage for convenience (only validation "train" examples in training, "test" excluded), but that shouldn't cause additional leakage compared to the two-staged train -> TTT (validation "train" examples are also included in second stage. Otherwise accuracy would be 100% with appropriate hyperparameter configuration, but we had never found across sweeps and Transformer baselines). 

Besides, thanks for the proposal! We're splitting the two stages, as "annealing" on the higher quality TTT-stage may improve performance. We welcome help and PR on this approach.

**Comment by imoneoi** [comment]
Btw I think there is the same misunderstanding in https://github.com/sapientinc/HRM/issues/1. As shown in https://github.com/sapientinc/HRM/blob/main/dataset/build_arc_dataset.py#L150-L151

"test" pairs / outputs in validation set are never used in the training / TTT stage, they're for evaluating performance. only "train" pairs in validation set are used for combined train+TTT which follows common practice.
